include/misc/sem.h

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * @brief public sys_sem APIs.
+ */
+
+#ifndef ZEPHYR_INCLUDE_MISC_SEM_H_
+#define ZEPHYR_INCLUDE_MISC_SEM_H_
+
+/*
+ * sys_sem behaves almost exactly like k_sem, with the added
+ * advantage that sys_sem instance can reside in user memory.
+ */
+
+#ifdef CONFIG_USERSPACE
+#include <kernel.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * sys_sem structure
+ */
+struct sys_sem {
+	struct k_futex futex;
+	/* maximum permitted sys_sem count.
+	 * but doesn't use it, put here to
+	 * make interface unify.
+	 */
+	atomic_val_t limit;
+};
+
+/**
+ * @brief Initialize a sys_sem.
+ *
+ * This routine initializes a sys_sem instance, prior to its first use.
+ *
+ * @param sem Address of the sys_sem.
+ * @param initial_count Initial sys_sem count.
+ * @param limit Maximum permitted futex val, not use it.
+ *
+ * @return N/A
+ */
+void sys_sem_init(struct sys_sem *sem, int initial_count, unsigned int limit);
+
+/**
+ * @brief Give a sys_sem.
+ *
+ * This routine gives @a sem. The producer of sys_sem will increase one
+ * atomically at every semaphore give whether there are waiters in futex
+ * kernel wait_queue. And then it will do k_futex_wake unconditionally
+ * to wakeup the threads waiting on this sys_sem. There is no good way
+ * to know the waiting number which is in kernel side, so we have to do
+ * k_futex_wake at every give.
+ *
+ * @param sem Address of the sys_sem.
+ *
+ * @return N/A
+ */
+void sys_sem_give(struct sys_sem *sem);
+
+/**
+ * @brief Take a sys_sem.
+ *
+ * This routine takes @a sem. This fuction handles the synchronization
+ * for sys_tem. And the consumer of sys_sem will decrease one at every
+ * semaphore take if futex's val greater than zero.
+ * But if futex's val is zero, the consumer of sys_sem_take will sleep
+ * until producer's wakeup or timeout happens depended on the timeout
+ * parameter. When timeout happens, the take will return directly. And
+ * for K_FOREVER case, the caller should also do the decrease after
+ * the wakeup by producer with the help of futex.
+ *
+ * @param sem Address of the sys_sem.
+ * @param timeout Waiting period to take the sys_sem (in milliseconds),
+ *                or one of the special values K_NO_WAIT and K_FOREVER.
+ *
+ * @retval 0 sys_sem taken.
+ * @retval -EBUSY Parameter address not recognized by the kernel.
+ * @retval -EAGAIN Waiting period timed out.
+ */
+int sys_sem_take(struct sys_sem *sem, s32_t timeout);
+
+/**
+ * @brief Get sys_sem's futex.
+ *
+ * This routine returns the current futex's val of @a sem.
```

Comment:
This documentation should NEVER mention futexes.
Explain what the API does, how it behaves.
End users are not concerned with how it works internally.
You can literally just cut and paste the documentation for k_sem.